examples/custom_rule.jl

```diff
@@ -0,0 +1,189 @@
+    y.dval .= 2 .* x.val .* x.dval
+    return Duplicated(out, sum(y.dval)) 
+end
+
+# In the signature of our rule, we have made use of `Enzyme`'s activity annotations. Let's break down each one:
+# - the [`Const`](@ref) annotation on `f` indicates that we accept a function `f` that does not have a derivative component,
+#   which makes sense since `f` itself does not depend on any parameters.
+# - the [`Duplicated`](@ref) annotation given in the second argument annotates the return value of `f`. This means that
+#   our `forward` function should return an output of type `Duplicated`, containing the original output `sum(y)` and its derivative.
+# - the [`Duplicated`](@ref) annotations for `x` and `y` mean that our `forward` function handles inputs `x` and `y`
+#   which have been marked as `Duplicated`. We should update their shadows with their derivative contributions. 
+
+# In the logic of our forward function, we run the original function, populate `y.dval` (the shadow of `y`), 
+# and finally return a `Duplicated` for the output as promised. Let's see our rule in action! 
+# With the same setup as before:
+
+x  = [3.0, 1.0]
+dx = [1.0, 0.0]
+y  = [0.0, 0.0]
+dy = [0.0, 0.0]
+
+g(y, x) = f(y, x)^2 # function to differentiate
+
+@show autodiff(Forward, g, Duplicated(y, dy), Duplicated(x, dx)) # derivative of g w.r.t. x[1]
+@show dy; # derivative of y w.r.t. x[1] when g is run
+
+# We see that our custom forward rule has been triggered and gives the same answer as before.
+
+# ## Handling more activities 
+
+# Our custom rule applies for the specific set of activities that are annotated for `f` in the above `autodiff` call. 
+# However, Enzyme has a number of other annotations. Let us consider a particular case as an example, where the output
+# has a [`DuplicatedNoNeed`](@ref) annotation. This means we are only interested in its derivative, not its value.
+# To squeeze the last drop of performance, the below rule avoids computing the output of the original function and 
+# just computes its derivative.
+
+function forward(func::Const{typeof(f)}, ::Type{<:DuplicatedNoNeed}, y::Duplicated, x::Duplicated)
+    println("Using custom rule with DuplicatedNoNeed output.")
+    y.val .= x.val.^2 
+    y.dval .= 2 .* x.val .* x.dval
+    return sum(y.dval)
+end
+
+# Our rule is triggered, for example, when we call `autodiff` directly on `f`, as the output's derivative isn't needed:
+
+x  = [3.0, 1.0]
+dx = [1.0, 0.0]
+y  = [0.0, 0.0]
+dy = [0.0, 0.0]
+
+@show autodiff(Forward, f, Duplicated(y, dy), Duplicated(x, dx)) # derivative of f w.r.t. x[1]
+@show dy; # derivative of y w.r.t. x[1] when f is run
+
+# !!! note "Custom rule dispatch"
+#     When multiple custom rules for a function are defined, the correct rule is chosen chosen using 
+#     [Julia's multiple dispatch](https://docs.julialang.org/en/v1/manual/methods/#Methods).
+#     In particular, it is important to understand that the custom rule does not *determine* the
```

Comment:
Is this correct? Was inspired by #695.

(And now that I think of it, if it's true, maybe there are some cases where we'd want custom activity annotations... e.g. if one writes `x, y = f(a, b)` where `f` is some function not understandable by Enzyme, and say the Jacobian of `f` is actually diagonal, it could be good to know that for activity analysis?)

Reply:
Yes, it is correct. The one thing to add is "If a custom rule is specified for the correct function/argument types, but not the correct activity annotation, a runtime error will be thrown alerting the user to the missing activity rule rather than silently ignoring the rule."